### PR TITLE
SCIM2 outbound provisioning connector.

### PR DESCRIPTION
**Pull request by kesavany**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+            scimProvider.setProperty(scimPropertyName, property.getDefaultValue());
+        }
+    }
+
+    @Override
+    public String getClaimDialectUri() throws IdentityProvisioningException {
+        return SCIM2ProvisioningConnectorConstants.DEFAULT_SCIM_DIALECT;
+    }
+
+    private void setUserPassword(User user, ProvisioningEntity userEntity) throws CharonException, BadRequestException {
```
Comment (menakaj) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: Shall we add Javadoc comments here? Also check the character length as well.

Comment (menakaj) [reply components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: Add comments to the other methods also.

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+    }
+
+    /*
+     * Construct the SCIM Object given the attribute URIs and attribute values of the object.
+     *
+     * @param attributes
+     * @param scimObjectType
+     * @return
+     */
+    public static SCIMObject constructSCIMObjectFromAttributes(Map<String, String> attributes,
```
Comment (menakaj) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java]: Formatting issue.

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+            case SCIM2CommonConstants.USER:
+                scimObject = new User();
+                if (debug) {
+                    log.debug("Building User Object");
+                }
+                break;
+            default:
+                break;
+        }
+        for (Map.Entry<String, String> attributeEntry : attributes.entrySet()) {
```
Comment (menakaj) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java]: Formatting issue. Better adding a new line to separate the switch and for.

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+        }
+    }
+
+    /*
+     * set claim mapping for multivalued attribute
+     *
+     * @param attribute
+     * @param claimsMap
+     */
+    private static void setClaimsForMultivaluedAttribute(Attribute attribute, Map<String, String> claimsMap) throws
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java]: Add a new line after method signature.

**Comment by malithie** [comment]
In addition, please add unit tests and integration tests and link with this PR.

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+            }
+        }
+    }
+
+    /**
+     * Initiates the SCIM2 operation.
+     *
+     * @param provisioningEntity
+     * @return provisionedEntity
+     * @throws IdentityProvisioningException
```
Comment (Megala21) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: Param description and Exception description is missing in java docs

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+            if (CollectionUtils.isNotEmpty(userNames)) {
+                userName = userNames.get(0);
+            }
+            User user;
+            // get single-valued claims
+            Map<String, String> singleValued = getSingleValuedClaims(userEntity.getAttributes());
+            // if user created through management console, claim values are not present.
+            if (MapUtils.isNotEmpty(singleValued)) {
+                user = (User) SCIMClaimResolver.constructSCIMObjectFromAttributes(singleValued,
+                        1);
```
Comment (Megala21) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: Seems to be a formatting issue

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java [hunk 7]
```diff
@@ -0,0 +1,609 @@
...
+        String immediateParentAttributeName = attributeNames[1];
+        String subAttributeURI = attributeEntry.getKey().replace("." + attributeNames[2], "");
+        String parentAttributeURI = subAttributeURI.replace(":" + attributeNames[1], "");
+        AttributeSchema subAttributeSchema = getAttributeSchema(subAttributeURI, scimObjectType);
+        AttributeSchema attributeSchema = getAttributeSchema(parentAttributeURI, scimObjectType);
+        // differentiate between sub attribute of Complex attribute and a Multivalued attribute with complex value
+        if (subAttributeSchema.getMultiValued()) {
+            SimpleAttribute typeSimpleAttribute = new SimpleAttribute(SCIMConstants.CommonSchemaConstants.TYPE,
+                    attributeNames[2]);
+            AttributeSchema typeAttributeSchema = getAttributeSchema(subAttributeSchema.getURI() + ".type",
```
Comment (Megala21) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java]: .type can we move to a constant, seems to be used in multiple places/

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java [hunk 8]
```diff
@@ -0,0 +1,609 @@
...
+                ComplexAttribute immediateParentAttribute = new ComplexAttribute(immediateParentAttributeName);
+                immediateParentAttribute.setSubAttribute(simpleAttribute);
+                DefaultAttributeFactory.createAttribute(subAttributeSchema, immediateParentAttribute);
+                // now super parent
+                AttributeSchema superParentAttributeSchema = getAttributeSchema(parentAttributeURI, scimObjectType);
+                ComplexAttribute superParentAttribute = new ComplexAttribute(superParentAttributeSchema.getName());
+                superParentAttribute.setSubAttribute(immediateParentAttribute);
+                DefaultAttributeFactory.createAttribute(superParentAttributeSchema, superParentAttribute);
+                // now add the super to the scim object
+                ((AbstractSCIMObject) scimObject).setAttribute(superParentAttribute);
```
Comment (Megala21) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/util/SCIMClaimResolver.java]: Shall we divide this to smaller methods?

File: features/etc/feature.properties [hunk 9]
```diff
@@ -0,0 +1,200 @@
+#
+# Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (Megala21) [inline features/etc/feature.properties]: License header should be 2018

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/pom.xml [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+							org.wso2.charon3.core.*; version="${charon.wso2.version.identity}",
+							org.wso2.scim2.*;version="${identity.scim2.client.version}",
+							io.scim2.swagger.client.*;version="${identity.scim2.client.version}"
+						</Import-Package>
+						<Export-Package>
+							!org.wso2.carbon.identity.provisioning.connector.scim2.internal,
+							org.wso2.carbon.identity.provisioning.connector.scim2.*;
+							version="${identity.outbound.provisioning.scim2.export.version}"
+						</Export-Package>
+						<DynamicImport-Package>*</DynamicImport-Package>
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/pom.xml]: Why dynamic import is needed ?
You should be able to remove this.

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+
+            if (provisioningEntity.getEntityType() == ProvisioningEntityType.USER) {
+                if (provisioningEntity.getOperation() == ProvisioningOperation.POST) {
+                    createUser(provisioningEntity);
+                } else if (provisioningEntity.getOperation() == ProvisioningOperation.DELETE) {
+                    deleteUser(provisioningEntity);
+                } else if (provisioningEntity.getOperation() == ProvisioningOperation.PUT) {
+                    updateUser(provisioningEntity, ProvisioningOperation.PUT);
+                } else {
+                    log.warn("Unsupported provisioning operation : " + provisioningEntity.getOperation());
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: log both the entity and the operation type

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+                }
+            } else if (provisioningEntity.getEntityType() == ProvisioningEntityType.GROUP) {
+                if (provisioningEntity.getOperation() == ProvisioningOperation.DELETE) {
+                    deleteGroup(provisioningEntity);
+                } else if (provisioningEntity.getOperation() == ProvisioningOperation.POST) {
+                    createGroup(provisioningEntity);
+                } else if (provisioningEntity.getOperation() == ProvisioningOperation.PUT) {
+                    updateGroup(provisioningEntity);
+                } else {
+                    log.warn("Unsupported provisioning operation : " + provisioningEntity.getOperation());
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: log both the entity and the operation type

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+                }
+            } else {
+                log.warn("Unsupported provisioning entity : " + provisioningEntity.getEntityName());
+            }
+        }
+        return null;
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: provisioning  the user and the group can be separated out for two methods improving readability

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+        try {
+            List<String> groupNames = getGroupNames(groupEntity.getAttributes());
+
+            if (CollectionUtils.isNotEmpty(groupNames)) {
+                groupName = groupNames.get(0);
+            }
+            Group group = new Group();
+            group.setDisplayName(groupName);
+            List<String> userList = getUserNames(groupEntity.getAttributes());
+            this.setGroupMembers(group, userList);
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: Don't invoke as this.setGroupMembers() here

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+        try {
+            List<String> groupNames = getGroupNames(groupEntity.getAttributes());
+            String groupName = null;
+            if (CollectionUtils.isNotEmpty(groupNames)) {
+                groupName = groupNames.get(0);
+            }
+            Group group = new Group();
+            group.setDisplayName(groupName);
+            List<String> userList = getUserNames(groupEntity.getAttributes());
+            this.setGroupMembers(group, userList);
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: No need of invoking as this. setGroupMembers

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+     * @param user
+     * @param userEntity
+     * @throws CharonException
+     * @throws BadRequestException
+     */
+    private void setUserPassword(User user, ProvisioningEntity userEntity) throws CharonException, BadRequestException {
+
+        if ("true".equals(scimProvider.getProperty(SCIM2ProvisioningConnectorConstants.
+                SCIM_ENABLE_PASSWORD_PROVISIONING))) {
+            this.setPassword(user, getPassword(userEntity.getAttributes()));
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: No need of invoking the private method as this.xxx

File: components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java [hunk 17]
```diff
@@ -0,0 +1,464 @@
...
+        } else if (StringUtils.isNotBlank(scimProvider.getProperty(SCIM2ProvisioningConnectorConstants.
+                SCIM_DEFAULT_PASSWORD))) {
+            this.setPassword(user, scimProvider.getProperty(SCIM2ProvisioningConnectorConstants.SCIM_DEFAULT_PASSWORD));
```
Comment (malithie) [inline components/org.wso2.carbon.identity.provisioning.connector.scim2/src/main/java/org/wso2/carbon/identity/provisioning/connector/scim2/SCIM2ProvisioningConnector.java]: No need of invoking the private method as this.xxx

**Pull request merged**